ENTESB-11495: DynamoDB Connector - Basic Funcionality

Delawen opened the pull request:
This PR introduces a new functionality using the AWS DynamodDB connector of Apache Camel. A new connector type is available (AWS-DDB) and two operations can be performed with that connector: PutItem and QueryItem.

To test this PR, create a new connection of type AWS-DDB:
![imagen](https://user-images.githubusercontent.com/726590/63429429-4a85fe80-c41a-11e9-83cb-ff597d25c39c.png)

That connection has to be configured with the proper credentials, which can be defined in [the Amazon AWS Console](https://console.aws.amazon.com/iam/home?region=us-east-2#/security_credentials) :
![imagen](https://user-images.githubusercontent.com/726590/63429491-7bfeca00-c41a-11e9-8138-37dac952e1d4.png)

Those credentials should be connected to some existing DynamoDB table, which can be created here: https://us-east-2.console.aws.amazon.com/dynamodb/home 
Make sure you create the table on the same region you configured on the connector. 

On "Table Name" we write the table name.

On my case, I created a table with a single ID (clave) of type string:
![imagen](https://user-images.githubusercontent.com/726590/63429664-fe878980-c41a-11e9-8282-f27b7a041aec.png)

Now we can create a simple integration, like a time based one to query an element:
![imagen](https://user-images.githubusercontent.com/726590/63449999-618c1700-c441-11e9-96ca-3cda33f7e516.png)

On "Filter" we define a json with all the constraints to extract elements (on our case, just match key column "clave").
On "Attributes to query" we optionally define a comma separated list of columns to extract (if empty, chooses the same columns as in the filter).

If you want to use the output for some later step you have to manually define the JSON schema:
![imagen](https://user-images.githubusercontent.com/726590/63429887-94bbaf80-c41b-11e9-8f71-19328132b9cb.png)

This JSON schema will be automatically added on later iterations. You can leave it empty.

The same way we can do a Put Item integration to insert elements on the database:

![imagen](https://user-images.githubusercontent.com/726590/63449905-3acde080-c441-11e9-8c7e-d1ecb964d0e4.png)

Delawen commented:
TODO/Currently dealing with: The current version of the Apache Camel connector forces me to set the tableName on the Connector configuration instead of on the operation:
![imagen](https://user-images.githubusercontent.com/726590/63448388-18869380-c43e-11e9-9f0c-551116cf4afe.png)

Quick fix: I am going to move the tableName config there, but it would be better to have it on the integration to have an auto-populated field.

heiko-braun commented:
+1 to the quick fix

Delawen commented:
I was a bit confused and had to do some force-pushes because I had locally built and installed the latest master version Apache Camel connector, which doesn't seem to match the one used on Syndesis by default :thinking: . So when upgrade the version of the Camel connectors in the pom.xml files, the verifier would allow us to move the table name to the operations.

oscerd commented:
The camel used by syndesis is in jboss-fuse/camel repository on the 2.21.x.redhat-7-x branch. Updating the camel version requires a new fuse 7.5 dev build and pinning the artifact. Syndesis actually doesn't use camel upstream

Delawen commented:
One of the checks failing because:

> [ERROR] Failed to execute goal on project connector-support-catalog: Could not resolve dependencies for project io.syndesis.connector:connector-support-catalog:jar:1.8-SNAPSHOT: Failed to collect dependencies at io.syndesis.connector:connector-aws-ddb:jar:1.8-SNAPSHOT: Failed to read artifact descriptor for io.syndesis.connector:connector-aws-ddb:jar:1.8-SNAPSHOT: Could not transfer artifact io.syndesis.connector:connector-aws-ddb:pom:1.8-SNAPSHOT from/to redhat-ga (https://maven.repository.redhat.com/ga/): Failed to transfer file: https://maven.repository.redhat.com/ga/io/syndesis/connector/connector-aws-ddb/1.8-SNAPSHOT/connector-aws-ddb-1.8-SNAPSHOT.pom. Return code is: 500 , ReasonPhrase:Internal Server Error. -> [Help 1]

Not sure if this is a problem on my side or if someone should create something on that repository as the connector is new.

squakez commented:
@Delawen you need to add your connector to the `cicd` as well in order to be picked up by the new parallel connectors deployment model: https://github.com/syndesisio/syndesis/blob/master/.circleci/config.yml#L964

Delawen commented:
Thanks! Let's see if it works now :)

squakez commented on app/connector/aws-ddb/src/test/java/io/syndesis/connector/aws/ddb/AWSDDBInsertItemTest.java:
```diff
@@ -0,0 +1,36 @@
+import org.junit.Ignore;
+
+@SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")
+@Ignore("Make sure the AWSDDBConfiguration has the proper credentials before running this test")
```
You should provide unit test with some embedded server. I had a quick look and it seems there are useful libraries for that: https://github.com/aws-samples/aws-dynamodb-examples/blob/master/src/test/java/com/amazonaws/services/dynamodbv2/DynamoDBLocalFixture.java - give it a try for next development iteration

squakez commented on app/connector/aws-ddb/src/main/java/io/syndesis/connector/aws/ddb/customizer/DDBConnectorCustomizer.java:
```diff
@@ -0,0 +1,139 @@
+
+            if (!element.isEmpty()) {
+                try {
+                    ObjectMapper mapper = new ObjectMapper();
```
Better make it instance variable or even static. ObjectMapper is thread safe.

Delawen replied:
Ok. I tend to avoid static as possible to save memory, but makes sense.

squakez commented on app/connector/aws-ddb/src/main/java/io/syndesis/connector/aws/ddb/customizer/DDBConnectorCustomizer.java:
```diff
@@ -0,0 +1,139 @@
+
+        if (options.containsKey(parameterName)) {
+
+            String element = options.get(parameterName).toString();
```
We have a utility method that can be safer: https://github.com/syndesisio/syndesis/blob/master/app/connector/support/util/src/main/java/io/syndesis/connector/support/util/ConnectorOptions.java#L40

squakez commented on app/connector/aws-ddb/src/main/java/io/syndesis/connector/aws/ddb/customizer/DDBConnectorCustomizer.java:
```diff
@@ -0,0 +1,139 @@
+                    for (Map.Entry<String, Object> attribute : attributes.entrySet()) {
+                        attributeMap.put(attribute.getKey(),
+                                new AttributeValue(attribute.getValue().toString()));
+                        if(LOG.isTraceEnabled()) {
```
Personally I don't like log checking, I think it pollutes the readability without any real gain

squakez commented on app/connector/aws-ddb/src/main/java/io/syndesis/connector/aws/ddb/customizer/DDBConnectorCustomizer.java:
```diff
@@ -0,0 +1,139 @@
+                    Map<String, Object> attributes = mapper.readValue(element, Map.class);
+
+                    if (attributes.isEmpty() && LOG.isWarnEnabled()) {
+                        LOG.warn("The parameter " + parameterName + " is an empty map.");
```
Improve it by `LOG.warn("The parameter {} is an empty map.", parameterName)`

squakez commented on app/connector/aws-ddb/src/main/java/io/syndesis/connector/aws/ddb/customizer/DDBConnectorCustomizerQuery.java:
```diff
@@ -0,0 +1,57 @@
+
+        List<String> attributes = new ArrayList<String>();
+        if (options.containsKey("attributes")) {
+            String optionAttributes = options.get("attributes").toString();
```
ConnectorOptions.extractOption

squakez commented on app/connector/aws-ddb/src/main/resources/META-INF/syndesis/connector/aws-ddb.json:
```diff
@@ -0,0 +1,235 @@
+          "kind": "any"
+        },
+        "outputDataShape": {
+          "kind": "any"
```
For a future development iteration, make sure to properly convert to the class provided by the component or to a json. Same apply to any kind of input/output data shape.

Delawen replied:
Yes, I am going to work next on extracting the table metadata once I finish with this PR.

squakez commented on app/connector/aws-ddb/src/main/resources/META-INF/syndesis/connector/verifier/aws-ddb:
```diff
@@ -0,0 +1,16 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+class=io.syndesis.connector.aws.ddb.verifier.DDBVerifier
```
Github does not like source not ended by blank space

squakez commented on app/connector/aws-ddb/src/test/java/io/syndesis/connector/aws/ddb/AWSDDBGenericOperation.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")
```
I don't think this is needed.

zregvart commented:
@Delawen do you plan on addressing some of the suggestions made in the review above or should we merge this and those will be addressed in subsequent pull requests?

Delawen commented:
> @Delawen do you plan on addressing some of the suggestions made in the review above or should we merge this and those will be addressed in subsequent pull requests?

I can work on this from Thursday on.

Delawen commented on app/connector/aws-ddb/pom.xml:
```diff
@@ -0,0 +1,188 @@
+  <name>Connector :: AWS DDB</name>
+  <packaging>jar</packaging>
+
+  <dependencyManagement>
```
So yes, AWS-S3 for example has the same dependency management. Trying to cleanup it anyway.

Delawen commented:
Waiting for the checks to complete, but this should be ready now. 

Instead of using the mock DynamoDB database, which had some issues with the way connectors are created, I just check the customizer behaves as expected on the unit test.

heiko-braun commented:
@Delawen The bot could not transition the ticket automatically, please update this Jira ticket manually: https://issues.jboss.org/browse/ENTESB-11495

squakez commented on app/connector/aws-ddb/pom.xml:
```diff
@@ -0,0 +1,174 @@
+    <repository>
+      <id>dynamodblocal</id>
+      <name>AWS DynamoDB Local Release Repository</name>
+      <url>https://s3-us-west-2.amazonaws.com/dynamodb-local/release</url>
```
Any reason why we need to add this repo?

Delawen replied:
Same as the next, removed as no longer used. Not using the DDB dummy server.

squakez commented on app/pom.xml:
```diff
@@ -129,6 +129,8 @@
     <micrometer.version>1.2.0</micrometer.version>
     <teiid.version>12.2.1</teiid.version>
+    <aws-java-sdk-core.version>1.11.269</aws-java-sdk-core.version>
+    <aws-dynamodb-local.version>1.11.86</aws-dynamodb-local.version>
```
Would it make sense to align to same core version too?

Delawen replied:
Didn't work with same version but now I realized that I included this for the dummy DDB server which I no longer use for testing. So removed!
